ngraph/test/backend/unary_test.hpp

```diff
@@ -0,0 +1,211 @@
+    //        generation and a static shape for final input tensor.
+    //
+    //  Shape{2, 4, 3}
+    //        A pure static shape for both ngraph function generation and input tensor
+    //
+    //  {}
+    //        A unspecified shape, test_unary() will derive the actual shape from the size
+    //        of the input vector as 1D shape.
+    struct ArgShape
+    {
+        PartialShape dynamic_shape;
+        Shape static_shape;
+        bool initialzed;
+        ArgShape(const PartialShape& dynamic_shape, const Shape& static_shape)
+            : dynamic_shape(dynamic_shape)
+            , static_shape(static_shape)
+            , initialzed(true)
+        {
+        }
+        ArgShape(const Shape& static_shape)
+            : dynamic_shape(static_shape)
+            , static_shape(static_shape)
+            , initialzed(true)
+        {
+        }
+        ArgShape()
+            : dynamic_shape{}
+            , static_shape{}
+            , initialzed(false)
+        {
+        }
+    };
+
+    template <typename TestEngine,
+              element::Type_t element_type,
+              typename value_type = fundamental_type_for<element_type>>
+    void test_unary(std::shared_ptr<Function> f,
+                    const std::vector<value_type>& test_input,
+                    const std::vector<value_type>& test_expected,
+                    ArgShape ashape = {},
+                    ArgTolerance tol = {})
+    {
+        if (!ashape.initialzed)
+        {
+            ashape = ArgShape(Shape({test_input.size()}));
+        }
+        // the size specified in static_shape may be bigger or smaller than the
+        // number of data contained in test_input & test_expected, here we
+        // sample test data repetitively to generate actual test data of required size.
+        auto static_size = shape_size(ashape.static_shape);
+        std::vector<value_type> input;
+        std::vector<value_type> expected;
+        for (size_t i = 0; i < static_size; i++)
+        {
+            input.push_back(test_input[i % test_input.size()]);
+            expected.push_back(test_expected[i % test_expected.size()]);
+        }
+
+        auto test_case = test::TestCase<TestEngine>(f);
+        test_case.template add_input<value_type>(input);
+        test_case.template add_expected_output<value_type>(ashape.static_shape, expected);
+        if (tol.using_bits)
+        {
+            if (tol.tolerance_bits >= 0)
+                test_case.run(tol.tolerance_bits);
+            else
+                test_case.run();
+        }
+        else
+        {
+            test_case.run_with_tolerance_as_fp(tol.tolerance_fp);
+        }
+    }
+
+    // overloading support passing reference function instead of expected results
+    template <typename TestEngine,
+              element::Type_t element_type,
+              typename value_type = fundamental_type_for<element_type>>
+    void test_unary(std::shared_ptr<Function> f,
```

Comment:
IMHO we should add function which return `expected` values from `inputs` and use only one `test_unary`

Reply:
That would require user to repeat the input vector again in it's code, which is not so good, right?
```

test_unary<TestEngine, element::f32>(
	create_unary_function<op::Acosh>(),
	{0.f, 1.f, -1.f, 2.f, -2.f, 3.f, -3.f, 4.f, 5.f, 10.f, 100.f},
	generate_expected({0.f, 1.f, -1.f, 2.f, -2.f, 3.f, -3.f, 4.f, 5.f, 10.f, 100.f}, std::acosh));

```

or it can be better:

```
const std::vector<float> input = {0.f, 1.f, -1.f, 2.f, -2.f, 3.f, -3.f, 4.f, 5.f, 10.f, 100.f};
auto expected = generate_expected(input, std::acosh)
test_unary<TestEngine, element::f32>(
	create_unary_function<op::Acosh>(),
	input,expected);
```
But caller still have to take extract care to use correct type for the input vector.

with this overload, one line code is possible and less cares & bugs from caller:
```
test_unary<TestEngine, element::f32>(
	create_unary_function<op::Acosh>(),
	{0.f, 1.f, -1.f, 2.f, -2.f, 3.f, -3.f, 4.f, 5.f, 10.f, 100.f},
	std::acosh);
```

Reply:
For me this approach is not flexible enough, but I will not persist.
```
template<elementType E, 
                typename Ret = std::vector<fundamental_type_for<E>>
                typenames ... Elements>
Ret  build_vec(Elements ... e){
    Ret  r;
    (void)std::initializer_list<int>{(r.push_back(e),0)...};
    return r;
}
```